.github/actions/build/action.yml

```diff
@@ -0,0 +1,211 @@
+      if: inputs.runs-on == 'macos-latest'
+      run: |
+        brew install flex bison openssl
+        echo 'ldflags=-l/usr/local/opt/openssl@3/lib' >> $GITHUB_ENV
+        echo 'cppflags=-i/usr/local/opt/openssl@3/include' >> $GITHUB_ENV
+        echo 'mold_run=' >> $GITHUB_ENV
+      shell: bash -euxo pipefail {0}
+
+    - name: Set pg 14 revision for caching
+      id: pg_v14_rev
+      run: echo ::set-output name=pg_rev::$(git rev-parse head:vendor/postgres-v14)
+      shell: bash -euxo pipefail {0}
+
+    - name: Set pg 15 revision for caching
+      id: pg_v15_rev
+      run: echo ::set-output name=pg_rev::$(git rev-parse head:vendor/postgres-v15)
+      shell: bash -euxo pipefail {0}
+
+    # Don't include the ~/.cargo/registry/src directory. It contains just
+    # uncompressed versions of the crates in ~/.cargo/registry/cache
+    # directory, and it's faster to let 'cargo' to rebuild it from the
+    # compressed crates.
+    - name: Cache cargo deps
+      id: cache_cargo
+      uses: actions/cache@v3
+      with:
+        path: |
+          ~/.cargo/registry/
+          !~/.cargo/registry/src
+          ~/.cargo/git/
+          target/
+        # Fall back to older versions of the key, if no cache for current Cargo.lock was found
+        key: |
+          v7-${{ runner.os }}-${{ inputs.build_type }}-cargo-${{ inputs.rust_toolchain }}-${{ hashfiles('cargo.lock') }}
+          v7-${{ runner.os }}-${{ inputs.build_type }}-cargo-${{ inputs.rust_toolchain }}-
+
+    - name: Cache postgres v14 build
+      id: cache_pg_14
+      uses: actions/cache@v3
+      with:
+        path: pg_install/v14
+        key: v1-${{ runner.os }}-${{ inputs.build_type }}-pg-${{ steps.pg_v14_rev.outputs.pg_rev }}-${{ hashFiles('Makefile') }}
+
+    - name: Cache postgres v15 build
+      id: cache_pg_15
+      uses: actions/cache@v3
+      with:
+        path: pg_install/v15
+        key: v1-${{ runner.os }}-${{ inputs.build_type }}-pg-${{ steps.pg_v15_rev.outputs.pg_rev }}-${{ hashFiles('Makefile') }}
+
+    - name: Build postgres v14
+      if: steps.cache_pg_14.outputs.cache-hit != 'true'
+      run: ${mold_run} make postgres-v14
+      shell: bash -euxo pipefail {0}
+
+    - name: Build postgres v15
+      if: steps.cache_pg_15.outputs.cache-hit != 'true'
+      run: ${mold_run} make postgres-v15
+      shell: bash -euxo pipefail {0}
+
+    - name: Build neon extensions
+      run: ${mold_run} make neon-pg-ext
+      shell: bash -euxo pipefail {0}
+
+    - name: Check formatting of Rust code
+      if: inputs.build_type == 'debug'
+      run: cargo fmt --all -- --check
+
+    - name: Run cargo build
+      run: |
+        ${cov_prefix} ${mold_run} cargo build $CARGO_FLAGS --features failpoints --bins --tests
+      shell: bash -euxo pipefail {0}
+
+    - name: Run cargo test
+      run: |
+        ${cov_prefix} cargo test $CARGO_FLAGS
+      shell: bash -euxo pipefail {0}
+
+    - name: Run cargo clippy
+      if: inputs.build_type == 'debug'
```

Comment:
IMO, we should style check `release` builds, as they are faster to run generally, here's a [recent](https://github.com/neondatabase/neon/actions/runs/3010591236) run:
<img width="478" alt="image" src="https://user-images.githubusercontent.com/2690773/189114282-b5e81c42-9591-401e-b133-a3b7dbb0bfb1.png">

So, if we add more to `debug` run, every PR would become slower to run. 
If we add to `release`, overall, the PR time might not even change (given that it finishes in ~8 minutes).

Also, with our current testing, slow `debug` builds are more useful when it comes to detection odd race issues and bad algorithms (e.g. walreceiver node selection).

If we add style checks to `debug` build, a person could spend a lot of time, rerunning debug builds to fix first the style checks, then all failing Python tests, with every `debug` run being quite slow now.
Adding style checks to `release` would allow to check both flacky tests with `debug` and style with `release` in a single CI run.

----------

It also feels more appropriate to have a separate parameter for style checks, instead of hardcoding them to run on a certain configuration. We could add a default value to such parameter, to make it more ergonomic.